redisson/src/main/java/org/redisson/connection/balancer/LoadBalancerManager.java

```diff
@@ -138,12 +138,18 @@ public boolean unfreeze(ClientConnectionsEntry entry, FreezeReason freezeReason)
 
             if (freezeReason != FreezeReason.RECONNECT
                     || entry.getFreezeReason() == FreezeReason.RECONNECT) {
-                entry.resetFirstFail();
-                entry.setFreezeReason(null);
-                
-                slaveConnectionPool.initConnections(entry);
-                pubSubConnectionPool.initConnections(entry);
-                return true;
+                if (!entry.isIniting()) {
+                    entry.setIniting(true);
+                    entry.resetFirstFail();
+                    slaveConnectionPool.initConnections(entry).onComplete((r, ex) -> {
```

Comment:
please use AsyncCountDownLatch here. Both handlers slaveConnectionPool, pubSubConnectionPool invoke countDown method and in latch callback we do rest logic.